feat(AP): rewrite objects matching algo and ap calculation

vividf opened the pull request:
Our previous approach used a different method for calculating Average Precision (AP) compared to the nuScenes AP metric.
To ensure that AWML adopts the same evaluation standard as this repository, we need to update the AP computation method and apply the `AP101` protocol.

To support this, this PR introduces a new class, `PerceptionFrame`, which allows `AWML` to store both ground truth and model predictions.  The reason why the new class is needed is that the `PerceptionFrameResult` includes too much unneeded information for only passing the objects from `AWML` to `autoware_perception_evaluation`.


In addition, `PerceptionFrameResult` now includes a new variable, `nuscene_object_results`, which organizes object results by (matching_mode, threshold). The reason why we couldn't use original object results to calcuate nusecne AP is because we match it with the closest distance in the first step and didn't consider the confidence score of each objects. 

Note that this change currently only affects the detection evaluation. In the future, we can also adopt nuscene_object_results as the official `object_results` for classification, tracking, and other tasks.


PR **before**+ AWML https://github.com/tier4/AWML/tree/feat/redesign_evaluation_pipeline
* boxes in ego coordinate

|      Label      | car(0.5) | truck(0.5) | bus(0.5) | bicycle(0.5) | pedestrian(0.5) |
|:---------------:|:------------:|:------------:|:------------:|:------------:|:------------:|
|   Predict_num   |     2949     |     112      |     185      |      7       |     768      |
|       AP        |    0.833     |    0.832     |    0.841     |    0.000     |    0.482     |


PR **after** + AWML https://github.com/tier4/AWML/tree/feat/redesign_evaluation_pipeline
* boxes in ego coordinate

|      Label      | car(0.5) | truck(0.5) | bus(0.5) | bicycle(0.5) | pedestrian(0.5) |
|:---------------:|:------------:|:------------:|:------------:|:------------:|:------------:|
|   Predict_num   |     2949     |     112      |     185      |      7       |     768      |
|       AP        |    0.803     |    0.808     |    0.815     |    0.000     |    0.389     |


AWML main branch 
* boxes in global coordinate
* Nuscene AP

|      Label      | car(0.5) | truck(0.5) | bus(0.5) | bicycle(0.5) | pedestrian(0.5) |
|:---------------:|:------------:|:------------:|:------------:|:------------:|:------------:|
|   Predict_num   |     2949     |     112      |     185      |      7       |     768      |
|       AP        |    0.804     |    0.808     |    0.815     |    0.000     |    0.389     |


As shown in the table above, this PR achieves almost the same AP as the current main branch of AWML.
The minor differences in the results can be attributed to the following reasons:

![awml_evaluation drawio](https://github.com/user-attachments/assets/f8baf63f-c782-40c8-ae53-343ad925c1c5)

`AWML` currently loads ground truth data directly from the T4Dataset rather than using the info pickle files.

Predicted objects undergo multiple coordinate transformations (global → ego → global) using `float32` precision, which introduces small numerical changes. These slight variations can affect the computed distances between objects, potentially causing a mismatch across threshold boundaries and leading to small differences in AP.

Ultimately, the AP computation we should follow is the one aligned with our current approach—evaluating all objects in the ego coordinate frame on a per-frame basis.

## Category

<!-- Please check an item that is most relative category to your changes. -->
<!-- Please delete options that are not relevant. -->

- [x] Perception
  - [x] Detection


## What

<!-- Please describe what you changed. -->

## Type of change


- [x] New feature


## Test performed


<!-- Describe how you have tested this PR. -->

- [ ] test.sensing_lsim

<details>
<pre>
<code>
<!-- Please paste test result here. -->
</code>
</pre>
</details>

- [ ] test.perception_lsim

<details>
<pre>
<code>
<!-- Please paste test result here. -->
</code>
</pre>
</details>

## Reference

<!-- Please write external reference if any. -->

## Notes for reviewer

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

KSeangTan commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -323,6 +324,104 @@ def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionRes
         )
 
 
+def get_nuscene_object_results(
```
Let's move this function to a new file instead

vividf replied:
I seperate them in two files in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/ff5395d41d62db9c0e8205160897d5a8f154110e

KSeangTan commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -323,6 +324,104 @@ def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionRes
+    if not estimated_objects:
+        return []
+
+    if metrics_config is None:
```
if `metrics_config` is not supposed to be None, then we shouldn't allow default value in the function definition

vividf replied:
True, addreseed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/0798b8715e6db38e0e9b6298ea2dce69853029cd

KSeangTan commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -323,6 +324,104 @@ def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionRes
+        MatchingMode.IOU3D: metrics_config.detection_config.iou_3d_thresholds,
+    }
+
+    object_results_dict: Dict[Tuple[str, float], List[DynamicObjectWithPerceptionResult]] = {}
```
Would you mind reminding me why the key can be either `str` or `float`?

vividf replied:
Thanks for the catch! It is a mistake after several refactorings! should be MatchingMode.
fixed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/2e6e2bc27417ba4641fc8fd8d80ed43a767198e3

KSeangTan commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -323,6 +324,104 @@ def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionRes
+
+    for matching_mode, threshold_list in matching_config_map.items():
+        if not threshold_list:
+            continue  # skip if config is None
```
Maybe it's better we move this checking to the line 363

vividf replied:
addressed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/2bba7de1780e092f6278250035b7a5b67e034bcf

KSeangTan reviewed (changes requested):
Thanks for the hard work, it looks good to me overall
Would you mind moving all functions related to `nuscenes` to new functions? It will be cleaner and easy for next refactoring

vividf commented on docs/ja/perception/design.md:
```diff
@@ -257,7 +257,7 @@ json_result = json.dump(dict_result)
 
 - 注目物体を動的決定するためのインターフェイス．
-- `PerceptionEvaluationManger`の初期化時ではなく，各フレーム毎(=callback)に指定する．
+- `PerceptionEvaluationManager`の初期化時ではなく，各フレーム毎(=callback)に指定する．
```
Thanks, 

![image](https://github.com/user-attachments/assets/bc2f52ea-2df9-4733-b84a-db5ca469b6fb)
Since I’ve fixed this spelling error, I think it’s a good opportunity to correct the related spelling errors as well.

vividf commented on perception_eval/perception_eval/evaluation/metrics/metrics.py:
```diff
@@ -119,69 +123,58 @@ def num_ground_truth(self) -> int:
         self,
-        object_results: Dict[LabelType, List[DynamicObjectWithPerceptionResult]],
+        object_results: Union[
+            Dict[LabelType, Dict[Tuple[MatchingMode, float], List[DynamicObjectWithPerceptionResult]]],  # Single frame
```
The original implementation already handles both single-frame and multi-frame inputs.
What I did in this PR is to make that distinction more explicit, so it’s clearer that the function is handling different types of input. Otherwise, it made me confused in the beginning.

I completely agree that the current structure is too complex, and I plan to refactor it—just not in this PR. Let’s keep that for the next one to keep the changes focused.

vividf commented on perception_eval/perception_eval/evaluation/result/perception_frame_result.py:
```diff
@@ -64,6 +68,7 @@ class PerceptionFrameResult:
     def __init__(
         self,
         object_results: List[DynamicObjectWithPerceptionResult],
+        nuscene_object_results: Dict[Tuple[str, float], List[DynamicObjectWithPerceptionResult]],
```
This cannot be made optional until we fully replace `object_results` with `nuscene_object_results` for all tasks—classification, detection, tracking, etc.

In the current design, tracking evaluation also includes detection evaluation, which means nuscene_object_results is required.

However, we will not know which evaluation module is being run until `evaluate_frame`, which  `PerceptionFrameResult` has already been created.

vividf replied:
Thanks for the suggestion. Before, I was planning to remove the object_results for the next PR, so I saved it directly.
But it is also nice that we separate them into two methods now. 
All of the changes related to Nuscene object results are modified in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/0cc6dac12e73b286f533eeec99881bb8c1d78f38

vividf commented on perception_eval/perception_eval/evaluation/metrics/detection/ap.py:
```diff
@@ -55,305 +52,152 @@ class Ap:
-            ap = 0.75 * (0.75 - 0.25) + 1.0 * (0.25 - 0.0)
-               = 0.625
+        # Define uniformly spaced recall levels for interpolation (101 points)
+        recall_interp = np.linspace(0.0, 1.0, 101)
```
Thanks for the advice!
addressed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/0ee12bde85e54d29ae3658fd499ed0d8d08f3825

KSeangTan commented on perception_eval/perception_eval/manager/perception_evaluation_manager.py:
```diff
@@ -159,6 +170,17 @@ def _filter_objects(
             **self.filtering_params,
         )
 
+        nuscene_object_results: Dict[
```
Let's use `NamedTuple` instead of `Tuple[]` as the key

vividf replied:
Thanks for the advice!
Addressed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/8971c711012e87cbce45841a954f6a87ec20d280

vividf commented on perception_eval/perception_eval/evaluation/result/perception_frame.py:
```diff
@@ -0,0 +1,73 @@
+from perception_eval.common.object import DynamicObject
+
+
+class PerceptionFrame:
```
This one can only support saving 3D objects (pred, gt). 
I added TODO here if we also utilize this class for 2d detection evaluation :) 
https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/2a93d8baf37d884f18e20967efe4a2510a839cde

vividf commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -323,6 +324,104 @@ def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionRes
+                        best_dist = dist
+                        best_gt_idx = gt_idx
+
+                is_match = best_gt_idx is not None and best_dist < threshold
```
Thanks for the catch!
Fixed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/9d45f62b49546d538977e123222c904a5ecccfb8

vividf replied:
![image](https://github.com/user-attachments/assets/2da0377e-79f6-49a8-b008-2bc8f168a542)

vividf commented on perception_eval/perception_eval/evaluation/metrics/metrics.py:
```diff
@@ -119,69 +123,58 @@ def num_ground_truth(self) -> int:
+        for (matching_mode, threshold), label_results in results_by_match_config.items():
+            target_labels = list(label_results.keys())
+            num_gt_dict = {label: num_ground_truth.get(label, 0) for label in target_labels}
+            threshold_list = [threshold] * len(target_labels)
```
Fixed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/a7264ab3e7d3af9d203599ea361b9125613dbcff.

Since renaming the `matching_threshold_list` would be a big change, let's do it in next PR.

vividf commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -323,6 +324,104 @@ def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionRes
+    object_results_dict: Dict[Tuple[str, float], List[DynamicObjectWithPerceptionResult]] = {}
+    estimated_objects_sorted = sorted(estimated_objects, key=lambda x: x.semantic_score, reverse=True)
+
+    for matching_mode, threshold_list in matching_config_map.items():
```
Fixed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/d084bf44cbf0edf27b70e878e0a750a82abe8262

vividf commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -323,6 +324,104 @@ def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionRes
+            # Change it to [0.5, 1.0]
+            threshold = threshold[0]
+
+            for est_obj in estimated_objects_sorted:
```
added in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/1d03834b1d44d23018d4b1b99dff838358df548c

vividf commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -323,6 +324,104 @@ def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionRes
+
+        matching_method_module, _ = _get_matching_module(matching_mode)
+        for threshold in threshold_list:
+            object_results: List[DynamicObjectWithPerceptionResult] = []
```
Thanks!
https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/06164cfd3789404e90e04ad09c4536c1d4e02b5e

vividf commented on perception_eval/perception_eval/evaluation/metrics/detection/ap.py:
```diff
@@ -55,305 +52,152 @@ class Ap:
 
         return tp_list, fp_list
 
+    def get_precision_recall_list(self) -> Tuple[List[float], List[float]]:
```
fixed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/9276bdb9c364518d03042ee35827bcd5c6c5726f

vividf commented on perception_eval/perception_eval/evaluation/matching/objects_filter.py:
```diff
@@ -671,46 +671,89 @@ def _is_target_object(
         if target_labels is not None and label not in target_labels:
             if isinstance(obj, DynamicObjectWithPerceptionResult) and obj.ground_truth_object is not None:
                 label = obj.ground_truth_object.semantic_label.label
+                if label not in target_labels:
```
It is the label from the estimated object and the ground truth object.
Added more explanation in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/d73171316c59f190e45f83eef29d6324d33141bc

vividf commented on perception_eval/perception_eval/evaluation/matching/objects_filter.py:
```diff
@@ -671,46 +671,89 @@ def _is_target_object(
+                return None
+        return label
+
+    if isinstance(objects, dict):
```
Thanks, added in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/9a38798839e53cc0609f1f593653b21cd3caf757. 

Some refactoring for the unit test needs fixing for the next PR. Added a comment in test_map.py

vividf commented:
@KSeangTan 
Thank you for your detailed review and suggestions!
I’ve fixed and refactored many parts of the code — please take a look when you have time.

Some areas, such as the tests, documentation, still need to be updated due to the recent function changes.
I propose handling those in a follow-up PR to keep this one from growing even larger.

ktro2828 commented on perception_eval/perception_eval/evaluation/metrics/detection/ap.py:
```diff
@@ -32,6 +29,12 @@
 logger = getLogger(__name__)
 
 
+# Number of recall thresholds for AP interpolation.
+# 101-point interpolation is a standard method used in object detection.
+# It uses recall levels at 0.00, 0.01, ..., 1.00 inclusive.
+NUM_RECALL_POINTS = 101
```
How about defining `NUM_CALL_POINTS=101` as a class variable of `Ap`?

vividf replied:
Thanks for your advice!
Addressed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/322cf92004d360e79e2386791a5f6d6e26bad229

KSeangTan commented on perception_eval/perception_eval/evaluation/result/object_result_matching.py:
```diff
@@ -0,0 +1,536 @@
+        return []
+
+    # There is no GT and not FP validation (= all FP)
+    if not ground_truth_objects and evaluation_task.is_fp_validation() is False:
```
```
not evaluation_task.is_fp_validation()
```

vividf replied:
I think this part is a bit unrelated, as it was copied from earlier code, but I appreciate the suggestion!
It definitely helps make it cleaner.

https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/e73685cf04f78158796872da6835241deacaff26

KSeangTan reviewed (changes requested):
Thanks for the work, would you mind making changes again? After this, i believe it's good to merge

vividf commented on perception_eval/perception_eval/evaluation/result/object_result.py:
```diff
@@ -304,9 +296,10 @@ def serialization(self) -> Dict[str, Any]:
     def deserialization(cls, data: Dict[str, Any]) -> DynamicObjectWithPerceptionResult:
         """Deserialize the data to DynamicObjectWithPerceptionResult."""
-        if data["opbject_type"] == DynamicObject2D.__name__:
+        if data["opject_type"] == DynamicObject2D.__name__:
```
@KSeangTan 
Thank you, and apologies for initially fixing it the wrong way :joy: 
https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/1ad232a6c8478788e249be5c7999bb7713025fd0

vividf commented on perception_eval/perception_eval/evaluation/metrics/metrics.py:
```diff
@@ -119,69 +124,57 @@ def num_ground_truth(self) -> int:
     def evaluate_detection(
         self,
-        object_results: Dict[LabelType, List[DynamicObjectWithPerceptionResult]],
+        object_results: Union[
```
The reason I used a `Union` type is because, in the original design, the input already supported both `List[DynamicObjectWithPerceptionResult] `and `List[List[DynamicObjectWithPerceptionResult]]`—representing single-frame and multi-frame inputs, respectively. However, this was not explicitly reflected in the type annotations before, which is why I added it.

`List[DynamicObjectWithPerceptionResult]`: Represents a single frame containing multiple estimated objects.

`List[List[DynamicObjectWithPerceptionResult]]`: Represents a scene across multiple frames, each with their own list of estimated objects.

The need for supporting both formats arises from our use cases: we perform both single-frame evaluations and scene-level (multi-frame) evaluations using this function.

vividf replied:
If you think we should change this now then I think we have two possible design options:

Option 1:
Separate the evaluation into two functions—one for single-frame evaluation and another for multi-frame (scene-level) evaluation.

Pros: Clear separation of responsibilities and input types.
Cons: Introduces code duplication since much of the logic would overlap.

Option 2:
Flatten the multi-frame input before calling evaluate_detection(), so it only accepts List[DynamicObjectWithPerceptionResult].

Pros: Simplifies the interface of evaluate_detection().
Cons: Requires significant changes to get_scene_result(), which currently handles multi-frame inputs for classification, tracking, etc.

KSeangTan replied:
Make sense, we should definitely change in the next PR 

IMO, we should go for the combinations for both options 

Always keep `evaluate_detection` with list of result frames, and assign a scenario id to them. After finish evaluation, we should aggregate the scenario results in another function

vividf replied:
https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/cc1244ddd87f915480991ed3873d1bcd31613d45

vividf commented on perception_eval/perception_eval/evaluation/matching/matching_config.py:
```diff
@@ -0,0 +1,11 @@
+from .object_matching import MatchingMode
+
+
+class MatchingConfig(NamedTuple):
```
Thanks!
Added in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/4a76c35a22a4304520a6b91bb86b0a21c6369795

vividf commented on perception_eval/perception_eval/evaluation/result/object_result_matching.py:
```diff
@@ -0,0 +1,536 @@
+from perception_eval.evaluation.result.object_result import DynamicObjectWithPerceptionResult
+
+
+def get_nuscene_object_results(
```
Argee!
https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/9e7141eb7fc5681d4332b31c47936838f48bc600

vividf commented on perception_eval/perception_eval/evaluation/matching/objects_filter.py:
```diff
@@ -671,46 +672,113 @@ def _is_target_object(
+        if gt_label in target_labels:
+            return gt_label
+
+    return None
```
Sorry, could you please clarify what you mean in more detail? Thank you!

KSeangTan replied:
We dont need to specify `return None` 
only `return` is sufficient

vividf commented on perception_eval/perception_eval/evaluation/matching/objects_filter.py:
```diff
@@ -671,46 +672,113 @@ def _is_target_object(
+            label = resolve_label(obj, target_labels)
+            if label is None:
                 continue
+            if matching_config not in result[label]:
```
Thanks for the suggestion! Addressed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/97c82c3bd9b17b94cf1ae87477541f66a6eb26d3

vividf commented on perception_eval/perception_eval/evaluation/result/object_result_matching.py:
```diff
@@ -0,0 +1,536 @@
+        best_gt_idx: Optional[int] = None
+
+        for gt_idx, gt_obj in enumerate(ground_truth_objects):
+            if gt_idx in matched_gt_ids:
```
Added in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/92900f32b0f75eefe64535bc55bd99eab38621b3

vividf commented on perception_eval/perception_eval/evaluation/matching/objects_filter.py:
```diff
@@ -671,46 +672,113 @@ def _is_target_object(
+
+
+def divide_nuscene_objects_by_label(
+    objects: Dict[MatchingConfig, List[DynamicObjectWithPerceptionResult]],
```
Thanks for the suggestion!
I renames several variables and functions name in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/372f03782bb40a57477a9bfc875df5ce1cf67992

vividf commented on perception_eval/perception_eval/evaluation/metrics/metrics.py:
```diff
@@ -119,69 +124,57 @@ def num_ground_truth(self) -> int:
-                is_detection_2d=self.evaluation_task.is_2d(),
-            )
-            self.maps.append(map_)
+        # Group by (matching_mode_str, threshold): Dict[LabelType -> List[DynamicObjectWithPerceptionResult]]
```
Thanks!
Addressed in https://github.com/tier4/autoware_perception_evaluation/pull/212/commits/372f03782bb40a57477a9bfc875df5ce1cf67992

KSeangTan reviewed (approved):
Hi @vividf 

Thanks for the quick fixes, i would say it's good to merge after resolving all comments.
Would you mind listing down all TODOs, and we should discuss that in person for the entire refactoring and priorities

KSeangTan commented:
Hi @ktro2828 @MasatoSaeki 

Would you mind taking a quick look as well? Appreciate it

vividf commented:
@ktro2828 @MasatoSaeki @scepter914 
If you have any suggestions, feel free to share them! 
I’ll proceed with merging this for now.
